interfaces/prompting/patterns.go

```diff
@@ -0,0 +1,534 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package prompting
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"regexp"
+	"strings"
+
+	doublestar "github.com/bmatcuk/doublestar/v4"
+)
+
+var ErrNoPatterns = errors.New("no patterns given, cannot establish precedence")
+
+// The default previously-expanded prefixes to which new patterns or expanded
+// groups are concatenated. This must be a slice containing the empty string,
+// since at the beginning of the pattern, we have only one prefix to which to
+// concatenate, and that prefix is the empty string. Importantly, this cannot
+// be an empty slice, since concatenating every entry in an empty slice with
+// every entry in a slice of expanded patterns would again result in an empty
+// slice.
+var defaultPrefixes = []string{""}
+
+// Limit the number of groups which are allowed to occur in a pattern.
+// This number includes all groups, including those nested or in series.
+const maxGroupsInPattern = 10
+
+// ExpandPathPattern expands all groups in the given path pattern.
+//
+// Groups are enclosed by '{' '}'. Returns a list of all the expanded path
+// patterns, or an error if the given pattern is invalid.
+func ExpandPathPattern(pattern string) ([]string, error) {
+	seenGroups := 0
+	if len(pattern) == 0 {
+		return nil, fmt.Errorf(`invalid path pattern: pattern has length 0`)
+	}
+	if strings.HasSuffix(pattern, `\`) && !strings.HasSuffix(pattern, `\\`) {
+		return nil, fmt.Errorf(`invalid path pattern: trailing unescaped '\' character: %q`, pattern)
+	}
+	reader := strings.NewReader(pattern)
+	currPrefixes := defaultPrefixes
+	currLiteralStart := 0
+	for {
+		r, _, err := reader.ReadRune()
```

Comment:
If you use the 2nd return value, you can keep track of the offset withuot having to call `Seek`.